src/ai/backend/manager/api/session.py

```diff
@@ -2178,6 +2180,111 @@ async def shutdown_service(request: web.Request, params: Any) -> web.Response:
     return web.Response(status=204)
 
 
+async def find_dependency_sessions(
+    root_session_name_or_id: Union[uuid.UUID, str],
+    db_connection: SAConnection,
+    access_key: AccessKey,
+):
+    dependency_session_cache: Dict[str, List] = {}
+
+    async def _find_dependency_sessions(session_name_or_id: Union[uuid.UUID, str]):
+        session_infos = await match_session_ids(
+            session_name_or_id,
+            access_key=access_key,
+            db_connection=db_connection,
+        )
+
+        assert len(session_infos) >= 1, "session not found!"
+
+        session_id = str(session_infos[0].get("session_id"))
+        session_name = session_infos[0].get("session_name")
+
+        assert isinstance(session_id, get_args(Union[uuid.UUID, str]))
+        assert isinstance(session_name, str)
+
+        session_dependency_query = (
+            sa.select(
+                [
+                    session_dependencies.c.depends_on,
+                ]
+            )
+            .select_from(session_dependencies)
+            .where(session_dependencies.c.session_id == session_id)
+        )
+
+        kernel_query = (
+            sa.select(
+                [
+                    kernels.c.status,
+                    kernels.c.status_changed,
+                ]
+            )
+            .select_from(kernels)
+            .where(kernels.c.session_id == session_id)
+        )
+
+        session_dependency_query_result, kernel_query_result = list(
+            map(
+                lambda result: result.mappings().all(),
+                await asyncio.gather(
+                    db_connection.execute(session_dependency_query),
+                    db_connection.execute(kernel_query),
+                ),
+            )
+        )
+
+        dependency_session_ids = list(
+            map(lambda x: str(x.get("depends_on")), session_dependency_query_result)
+        )
```

Comment:
This will be much more neat.

```suggestion
        dependency_session_ids = [
            str(x.get("depends_on"))
            for x in session_dependency_query_result
        ]
```